src/dune_pkg/fiber_cache.ml

```diff
@@ -0,0 +1,43 @@
+open! Import
+
+type ('k, 'v) t =
+  { table : ('k, 'v Fiber.Ivar.t) Table.t
+  ; key_module : (module Table.Key with type t = 'k)
+  }
+
+let create key_module = { table = Table.create key_module 1; key_module }
+
+let find_or_add t key ~f =
+  let open Fiber.O in
+  let* () = Fiber.return () in
+  match Table.find t.table key with
+  | Some ivar -> Fiber.Ivar.read ivar
+  | None ->
+    let ivar = Fiber.Ivar.create () in
+    (* Add the empty ivar to the table before calling [f] so that if the key
+       requested again while [f] is running then the caller will wait on the
+       ivar that will eventually contain the result.
+
+       Note that this won't raise because [Table.find] and [Table.add_exn] are
+       called in the same fiber note so there's no possibility for a
+       conflicting entry to be added to the table in between those calls. *)
+    Table.add_exn t.table key ivar;
+    let* value = f () in
```

Comment:
To be extra safe, you can use `Fiber.with_error_handler` to evict `key` in case of an error. That way subsequent calls to the same will not deadlock.

Reply:
Updated to handle the case where [f] raises. Concurrent calls (while the ivar is present in the table) will raise, and subsequent calls (after the ivar is removed) will attempt to recompute the value.